# sRNAflow


## To do

### Shiny
* https://shiny.rstudio.com/gallery/widget-gallery.html

### Docker client for R https://richfitz.github.io/stevedore 

### NextFlow vs ...
* https://www.nextflow.io/docs/latest/index.html
* https://jmazz.me/blog/NGS-Workflows
* DolphinNext https://github.com/UMMS-Biocore/dolphinnext  https://www.biorxiv.org/content/10.1101/689539v3.full
* Awesome Nextflow https://github.com/nextflow-io/awesome-nextflow/

### metagenomeSeq
https://bioconductor.org/packages/release/bioc/html/metagenomeSeq.html

### MultiQC
https://multiqc.info/docs/#using-multiqc-in-pipelines

### Anton Enright Lab https://github.com/EnrightLab/Courses-and-Practicals/tree/master/Cambridge_PI 

### Clustering of genomes from the same spiece:

* clustering of small RNA sequences https://seqcluster.readthedocs.io/getting_started.html
* sumaclust https://git.metabarcoding.org/obitools/sumaclust/wikis/home


### Identifying significantly impacted pathways: a comprehensive review and assessment
* https://genomebiology.biomedcentral.com/articles/10.1186/s13059-019-1790-4?utm_source=Advaita+Bio+Newsletter&utm_campaign=74e7679cb2-genome-biology-benchmarking_2019-oct_resend&utm_medium=email&utm_term=0_8c276855a3-74e7679cb2-87484121 SPIA  https://doi.org/10.1186/s13059-019-1790-4


### Example files from:
* Incorporate dee2 data into your R-based RNA-seq workflow http://genomespot.blogspot.com/2018/09/incorporating-dee2-data-into-your-r.html

